cirq/contrib/two_qubit_gates/math_utils.py

```diff
@@ -0,0 +1,292 @@
+_negations[3, (0, 1, 2), (0, 1, 2)] = (-1, -1, 1)
+
+_offsets = np.zeros((4, 3))
+_offsets[1, (1, 2)] = np.pi / 2
+_offsets[1, (0, 2)] = np.pi / 2
+_offsets[1, (0, 1)] = np.pi / 2
+
+
+def _kak_equivalent_vectors(kak_vec) -> np.ndarray:
+    """Generates all KAK vectors equivalent under single qubit unitaries."""
+    # coverage: ignore
+    kak_vec = np.asarray(kak_vec)
+    # Produce all shift-negations of the kak vector
+    out = np.einsum('nab,...b->...na', _negations, kak_vec,
+                    dtype=float)  # (...,4,3)
+    out[..., :, :] += _offsets
+
+    # Apply all permutations of indices
+    out = np.einsum('pcb,...nb->...pnc', _perms_123, out)  # (...,6,4,3)
+
+    # Merge indices
+    return np.reshape(out, out.shape[:-3] + (24, 3))
+
+
+def KAK_vector_infidelity(k_vec_a: np.ndarray,
+                          k_vec_b: np.ndarray,
+                          ignore_equivalent_vectors: bool = False
+                         ) -> np.ndarray:
+    """Minimum entanglement infidelity between two KAK vectors. """
+
+    if ignore_equivalent_vectors:
+        k_diff = k_vec_a - k_vec_b
+        out = 1 - np.product(np.cos(k_diff), axis=-1)**2
+        out -= np.product(np.sin(k_diff), axis=-1)**2
+        return out
+    # coverage: ignore
+    # We must take the minimum infidelity over all possible locally equivalent
+    # KAK vectors. We need only consider equivalent vectors of one input.
+    k_vec_a = np.asarray(k_vec_a)[..., np.newaxis, :]  # (...,1,3)
+    k_vec_b = _kak_equivalent_vectors(np.asarray(k_vec_b))  # (...,24,3)
+
+    k_diff = k_vec_a - k_vec_b
+
+    out = 1 - np.product(np.cos(k_diff), axis=-1)**2
+    out -= np.product(np.sin(k_diff), axis=-1)**2  # (...,24)
+
+    return out.min(axis=-1)
+
+
+def in_weyl_chamber(xp: np.ndarray, yp: np.ndarray,
+                    zp: np.ndarray) -> np.ndarray:
+    """Whether a given collection of coordinates is within the Weyl chamber.
+
+    Args:
+        xp: X coordinates for the KAK vector.
+        yp: Y coordinates for the KAK vector. Must have same shape as xp.
+        zp: Z coordinates for the KAK vector. Must have same shape as xp.
+
+    Returns:
+        np.ndarray of boolean values denoting whether the given coordinates
+        are in the Weyl chamber.
+    """
+    pi_4 = np.pi / 4
+
+    x_inside = np.logical_and(0 <= xp, xp <= pi_4)
+
+    y_inside = np.logical_and(0 <= yp, yp <= pi_4)
+    y_inside = np.logical_and(y_inside, xp >= yp)
+
+    z_inside = np.abs(zp) <= yp
+
+    out = np.logical_and(x_inside, y_inside)
+    return np.logical_and(out, z_inside)
+
+
+def weyl_chamber_mesh(spacing: float) -> np.ndarray:
+    """Cubic mesh of points in the Weyl chamber.
+
+    Args:
+        spacing: Euclidean distance between neighboring KAK vectors.
+
+    Returns:
+        np.ndarray of shape (N,3) corresponding to the points in the Weyl
+        chamber.
+    """
+    if spacing < 1e-3:  # memory required ~ 1 GB
+        # coverage: ignore
```

Comment:
This actually seems like a good thing to test :)

Reply:
Done